src/array/cpu/id_hash_map.cc

```diff
@@ -0,0 +1,201 @@
+/**
+ *  Copyright (c) 2023 by Contributors
+ * @file array/cpu/id_hash_map.cc
+ * @brief Class about id hash map
+ */
+
+#include "id_hash_map.h"
+
+#ifdef _MSC_VER
+#include <intrin.h>
+#endif  // _MSC_VER
+
+#include <dgl/array.h>
+#include <dgl/runtime/parallel_for.h>
+
+#include <cmath>
+#include <numeric>
+
+using namespace dgl::runtime;
+
+namespace {
+static constexpr int64_t kEmptyKey = -1;
+static constexpr int kGrainSize = 256;
+
+// The formula is established from experience which is used
+// to get the hashmap size from the input array size.
+inline size_t GetMapSize(size_t num) {
+  size_t capcacity = 1;
+  return capcacity << static_cast<size_t>(1 + std::log2(num * 3));
+}
+}  // namespace
+
+namespace dgl {
+namespace aten {
+
+template <typename IdType>
+IdType IdHashMap<IdType>::CompareAndSwap(
+    IdType* ptr, IdType old_val, IdType new_val) {
+#ifdef _MSC_VER
+  if (sizeof(IdType) == 4) {
+    return _InterlockedCompareExchange(
+        reinterpret_cast<LONG*>(ptr), new_val, old_val);
+  } else if (sizeof(IdType) == 8) {
+    return _InterlockedCompareExchange64(
+        reinterpret_cast<LONGLONG*>(ptr), new_val, old_val);
+  } else {
+    LOG(FATAL) << "ID can only be int32 or int64";
+  }
+#elif __GNUC__  // _MSC_VER
+  return __sync_val_compare_and_swap(ptr, old_val, new_val);
+#else           // _MSC_VER
```

Comment:
Does that mean compare and swap is not supported on Mac (clang)?

Reply:
No，it is also supported in mac. And I have did some tests in mac and it works well.